server/auth/types/jwt/jwt_auth.ts

```diff
@@ -0,0 +1,145 @@
+/*
+ *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+import { SecurityPluginConfigType } from '../../..';
+import {
+  SessionStorageFactory,
+  IRouter,
+  IClusterClient,
+  CoreSetup,
+  AuthenticationHandler,
+  KibanaRequest,
+} from '../../../../../../src/core/server';
+import { SecuritySessionCookie } from '../../../session/security_cookie';
+import { SecurityClient } from '../../../backend/opendistro_security_client';
+import { User } from '../../user';
+import { parse, format } from 'url';
+import { get } from 'lodash';
+import { ParsedUrlQuery } from 'querystring';
+
+export class JwtAuthentication {
+  private static readonly AUTH_TYPE: string = 'jwt';
+  private authHeaderName: string;
+  private securityClient: SecurityClient;
+
+  constructor(
+    private readonly config: SecurityPluginConfigType,
+    private readonly sessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>,
+    private readonly router: IRouter,
+    private readonly esClient: IClusterClient,
+    private readonly coreSetup: CoreSetup
+  ) {
+    this.authHeaderName = this.config.jwt?.header.toLowerCase() || 'authorization';
+    this.securityClient = new SecurityClient(esClient);
+  }
+
+  authHandler: AuthenticationHandler = async (request, response, tookit) => {
```

Comment:
nonblocking comment: probably break this function into 3 subfunctions for better readability, authenticate by jwt, authenticate by cookie, or redirect.